File: solo-leveling-notification-service/src/main/java/com/sleepkqq/sololeveling/notification/kafka/SendNotificationConsumer.java
package com.sleepkqq.sololeveling.notification.kafka;

import static com.sleepkqq.sololeveling.avro.constants.KafkaGroupIds.NOTIFICATION_GROUP_ID;
import static com.sleepkqq.sololeveling.avro.constants.KafkaTaskTopics.SEND_NOTIFICATION_TOPIC;

import com.sleepkqq.sololeveling.avro.notification.ReceiveNotificationEvent;
import com.sleepkqq.sololeveling.avro.notification.SendNotificationEvent;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class SendNotificationConsumer {

  private final ReceiveNotificationProducer receiveNotificationProducer;

  @KafkaListener(topics = SEND_NOTIFICATION_TOPIC, groupId = NOTIFICATION_GROUP_ID)
  public void listen(SendNotificationEvent event) {
    log.info("<< Start sending notification | transactionId={}", event.getTransactionId());
    var receiveNotificationEvent = new ReceiveNotificationEvent(
        event.getTransactionId(),
        event.getUserId(),
        event.getNotification()
    );
    receiveNotificationProducer.send(event.getPriority(), receiveNotificationEvent);
  }
}

File: solo-leveling-notification-service/src/main/kotlin/com/sleepkqq/sololeveling/notification/kafka/SendNotificationConsumer.kt
package com.sleepkqq.sololeveling.notification.kafka

import com.sleepkqq.sololeveling.avro.config.consumer.AbstractKafkaConsumer
import com.sleepkqq.sololeveling.avro.constants.KafkaTaskTopics
import com.sleepkqq.sololeveling.avro.idempotency.IdempotencyService
import com.sleepkqq.sololeveling.avro.notification.SendNotificationEvent
import com.sleepkqq.sololeveling.notification.mapper.AvroMapper
import org.slf4j.LoggerFactory
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.kafka.annotation.RetryableTopic
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class SendNotificationConsumer(
	private val receiveNotificationProducer: ReceiveNotificationProducer,
	private val avroMapper: AvroMapper,
	idempotencyService: IdempotencyService
) : AbstractKafkaConsumer<SendNotificationEvent>(
	idempotencyService = idempotencyService,
	log = LoggerFactory.getLogger(SendNotificationConsumer::class.java)
) {

	@Transactional
	@RetryableTopic
	@KafkaListener(
		topics = [KafkaTaskTopics.UI_NOTIFICATION_TOPIC],
		groupId = $$"${spring.kafka.avro.group-id}"
	)
	fun listen(event: SendNotificationEvent) {
		consumeWithIdempotency(event)
	}

	override fun getTxId(event: SendNotificationEvent): String = event.txId

	override fun processEvent(event: SendNotificationEvent) {
		log.info("<< Start sending notification | txId={}", event.txId)
		val receiveNotificationEvent = avroMapper.map(event)
		receiveNotificationProducer.send(event.priority, receiveNotificationEvent)
	}
}
